### PR TITLE
Add customer page

### PR DESCRIPTION
**Pull request by sangvu**
Hi All,
Please help me review this PR. This include
 - GUI: Add customer page under http://localhost:8081/friend, make by UI Element framework
- Add, edit, update, delete customer APIs in backend

Thanks,
Sang Vu

### CODE REVIEW

File: internal/app/api/api.go [hunk 1]
```diff
@@ -74,6 +74,31 @@ func Init(conns *InfraConns) (http.Handler, error) {
...
+			method:  "GET",
+			handler: indexWebHandler.FindFriendByID,
+		},
+		{
+			path:    "/customer",
+			method:  "PUT",
+			handler: indexWebHandler.UpdateFriend,
+		},
+		{
+			path:    "/customers",
```
Comment (gaconkzk) [inline internal/app/api/api.go]: for consistency, use `customers` for all CRUD path.

File: internal/app/api/api.go [hunk 2]
```diff
@@ -74,6 +74,31 @@ func Init(conns *InfraConns) (http.Handler, error) {
...
+			method:  "GET",
+			handler: indexWebHandler.GetAllFriends,
+		},
+		{
+			path:    "/customer",
```
Comment (gaconkzk) [inline internal/app/api/api.go]: which customer that we want to delete? missing {id}

File: internal/app/api/api.go [hunk 3]
```diff
@@ -74,6 +74,31 @@ func Init(conns *InfraConns) (http.Handler, error) {
...
+			method:  "POST",
+			handler: indexWebHandler.AddFriend,
+		},
+		{
+			path:    "/customer/{id}",
+			method:  "GET",
+			handler: indexWebHandler.FindFriendByID,
+		},
+		{
+			path:    "/customer",
```
Comment (pthethanh) [inline internal/app/api/api.go]: should be:
GET /customers --> without body
DELETE, GET /customers/{id} --> without body
PUT /customers/{id} --> with body

"customers" should be "friends"?

File: internal/app/api/handler/index/index.go [hunk 4]
```diff
@@ -1,25 +1,33 @@
 package indexhandler
 
 import (
+	"fmt"
 	"io"
 	"net/http"
 	"os"
-
+	"encoding/json"
+	"github.com/globalsign/mgo/bson"
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: handler, service layer should not depend on a specific implementation of the repository. If you do this, then what if we change the repository to use a flat file, MySQL or something else,...?

File: internal/app/api/handler/index/index.go [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
+	"github.com/TheFlies/ofriends/internal/app/types"
+	"github.com/TheFlies/ofriends/internal/app/friend"
+	"github.com/gorilla/mux"
 )
 
 type (
 	// Handler is index web handler
-	Handler struct{}
+	Handler struct{
+		repo friend.Repository
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: Handler should use a service, not contact repo directly.

File: internal/app/api/handler/index/index.go [hunk 6]
```diff
@@ -30,3 +38,69 @@ func (web *Handler) Index(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func (web *Handler) AddFriend(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: what if the body is nil?

File: internal/app/api/handler/index/index.go [hunk 7]
```diff
@@ -30,3 +38,69 @@ func (web *Handler) Index(w http.ResponseWriter, r *http.Request) {
...
+		respond.Error(w, err, http.StatusBadRequest)
+		return
+	}
+	friend.Id = bson.NewObjectId()
+	if err := web.repo.InsertFriend(friend); err != nil {
+		respond.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	respond.JSON(w, http.StatusCreated, map[string]string{"result": "success"})
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: a POST should return id of the created object. It's useful for the UI to know the id to process next steps. Please check and fix all similar mistakes

File: internal/app/api/handler/index/index.go [hunk 8]
```diff
@@ -1,25 +1,33 @@
...
 // New return new index web handler
-func New() *Handler {
-	return &Handler{}
+func New(repo friend.Repository) *Handler {
+	return &Handler{repo}
 }
 
 // Index render index page
 func (web *Handler) Index(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintln(w, "Index!")
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: Is this a mistake? Please revert.

File: internal/app/api/handler/index/index.go [hunk 9]
```diff
@@ -30,3 +38,69 @@ func (web *Handler) Index(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func (web *Handler) AddFriend(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	var friend *types.Friend
+
+	if err := json.NewDecoder(r.Body).Decode(&friend); err != nil {
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: are u using pointer to a pointer? check and fix all same issues

File: internal/app/api/handler/index/index.go [hunk 10]
```diff
@@ -30,3 +38,69 @@ func (web *Handler) Index(w http.ResponseWriter, r *http.Request) {
...
+		respond.Error(w, err, http.StatusBadRequest)
+		return
+	}
+	friend.Id = bson.NewObjectId()
```
Comment (pthethanh) [inline internal/app/api/handler/index/index.go]: handler should know nothing about what kind of database to be used.

File: internal/app/friend/mongodb_repo.go [hunk 11]
```diff
@@ -22,17 +22,60 @@ func NewMongoRepository(s *mgo.Session) *MongoRepository {
...
 	}
 	return friend, nil
 }
 
+// FindByID return friend base on given id
+func (r *MongoRepository) FindFriendByID(id string) (*types.Friend, error) {
+	s := r.session.Clone()
+	defer s.Close()
+	var friend *types.Friend
+	if err := r.collection(s).Find(bson.M{"_id": bson.ObjectIdHex(id)}).One(&friend); err != nil {
```
Comment (pthethanh) [inline internal/app/friend/mongodb_repo.go]: pointer to a pointer?

File: internal/app/friend/mongodb_repo.go [hunk 12]
```diff
@@ -22,17 +22,60 @@ func NewMongoRepository(s *mgo.Session) *MongoRepository {
...
+	}
+	return friend, nil
+}
+
+
+// Insert Friend into database
+func (r *MongoRepository) InsertFriend(friend *types.Friend) error {
+	s := r.session.Clone()
+	defer s.Close()
+	err := r.collection(s).Insert(&friend)
```
Comment (pthethanh) [inline internal/app/friend/mongodb_repo.go]: return r.collection(s)....

File: internal/app/types/friend.go [hunk 13]
```diff
@@ -1,7 +1,16 @@
...
 
+import (
+	"github.com/globalsign/mgo/bson"
+)
+
 // Friend hold information of a friend
 type Friend struct {
-	ID   string `json:"id" bson:"id"`
-	Name string `json:"name"`
+	Id    bson.ObjectId `bson:"_id" json:"id"`
```
Comment (pthethanh) [inline internal/app/types/friend.go]: consider not to do this as your implementation assume that we will use mongodb for entire our life :)

File: internal/app/types/friend.go [hunk 14]
```diff
@@ -1,7 +1,16 @@
...
+	Name        string  `bson:"name" json:"name"`
+	Title       string  `bson:"title" json:"title"`
+	Project     string  `bson:"project" json:"project"`
+	Birthday    string  `bson:"birthday" json:"birthday"`
+	FoodNote    string  `bson:"foodnote" json:"foodnote"`
+	Family      string  `bson:"family" json:"family"`
```
Comment (pthethanh) [inline internal/app/types/friend.go]: How is about other information like: company, office that he/she is working, current position, country, city....?

File: internal/pkg/middleware/cors.go [hunk 15]
```diff
@@ -7,6 +7,12 @@ import (
 func EnableCORS(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.Header().Set("Access-Control-Allow-Origin", "*")
-		h.ServeHTTP(w, r)
+		w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT, DELETE")
```
Comment (pthethanh) [inline internal/pkg/middleware/cors.go:10]: If I were u, I would use `handlers.CORS()` (see github.com/gorilla/handlers)

File: web/src/App.vue [hunk 16]
```diff
@@ -1,26 +1,16 @@
...
-        If Element is successfully added to this project, you'll see an
-        <code v-text="'<el-button>'"></code>
-        below
-      </p>
-      <el-button>el-button</el-button>
-    </div>
-    <HelloWorld msg="Welcome to Your Vue.js App"/>
+    <router-link :to="{ name: 'Hello' }">Home</router-link>
+    <router-link to="/about">About</router-link>
+    <router-link to="/customer">customer</router-link>
```
Comment (pthethanh) [inline web/src/App.vue]: Please be consistency

File: web/src/components/AddCustomer.vue [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    <el-form ref="form" :model="form" :rules="rules" label-width="100px">
+      <h3>Add New Customer</h3>
+      <el-form-item label="Name" prop="name">
+        <el-input type="success" placeholder="Fistname and last name" v-model="form.name"></el-input>
+      </el-form-item>
+      <el-form-item label="Title" prop="title">
+        <el-input placeholder="VP" v-model="form.title"></el-input>
+      </el-form-item>
+      <el-form-item label="Project" prop="project">
+        <el-input placeholder="Accedian..." v-model="form.project"></el-input>
```
Comment (pthethanh) [inline web/src/components/AddCustomer.vue]: Please don't leak customer names

File: internal/app/api/api.go (hunk 1 above)
Comment (sangvu) [reply internal/app/api/api.go]: Fixed, I changed all to "friends" as aThanh's suggestion

File: internal/app/api/api.go (hunk 2 above)
Comment (sangvu) [reply internal/app/api/api.go]: Updated

File: internal/app/api/api.go (hunk 3 above)
Comment (sangvu) [reply internal/app/api/api.go]: Updated

File: internal/app/api/handler/index/index.go (hunk 4 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/api/handler/index/index.go (hunk 5 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: fixed

File: internal/app/api/handler/index/index.go (hunk 6 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/api/handler/index/index.go (hunk 7 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/api/handler/index/index.go (hunk 8 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/api/handler/index/index.go (hunk 9 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Sorry, I couldn't got your idea

File: internal/app/api/handler/index/index.go (hunk 10 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/api/handler/index/index.go (hunk 9 above)
Comment (sangvu) [reply internal/app/api/handler/index/index.go]: Fixed

File: internal/app/friend/mongodb_repo.go (hunk 11 above)
Comment (sangvu) [reply internal/app/friend/mongodb_repo.go]: Fixed

File: internal/app/types/friend.go (hunk 13 above)
Comment (sangvu) [reply internal/app/types/friend.go]: Fixed

File: internal/app/types/friend.go (hunk 14 above)
Comment (sangvu) [reply internal/app/types/friend.go]: Added

File: web/src/components/AddCustomer.vue (hunk 17 above)
Comment (sangvu) [reply web/src/components/AddCustomer.vue]: Fixed

File: internal/pkg/middleware/cors.go (hunk 15 above)
Comment (sangvu) [reply internal/pkg/middleware/cors.go:10]: package middleware


import (
	"net/http"

	"github.com/gorilla/handlers"
)

func EnableCORS(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		handlers.CORS()
	})
}
I tried but it's didn't work

File: web/src/App.vue (hunk 16 above)
Comment (sangvu) [reply web/src/App.vue]: fixed

File: internal/app/api/handler/friend/friend.go [hunk 18]
```diff
@@ -40,3 +47,75 @@ func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {
...
+	
+	respond.JSON(w, http.StatusOK, friends)
+
+}
+
+// Create handle insert gift HTTP Request
+func (h *Handler) Create(w http.ResponseWriter, r *http.Request) {
+	var friend types.Friend
+
+	if (r.Body == nil) {
```
Comment (pthethanh) [inline internal/app/api/handler/friend/friend.go]: delete this block of code. Move the line `defer r.Body.Close() ` to line #81

File: internal/app/api/handler/friend/friend.go [hunk 19]
```diff
@@ -40,3 +47,75 @@ func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {
...
+		}
+		respond.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	if err := h.srv.Create(r.Context(), friend); err != nil {
+		respond.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+	respond.JSON(w, http.StatusCreated, friend.ID)
```
Comment (pthethanh) [inline internal/app/api/handler/friend/friend.go]: should return `{"id":friend.ID}`

File: internal/app/api/handler/friend/friend.go [hunk 20]
```diff
@@ -40,3 +47,75 @@ func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {
...
+		}
+		respond.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	if err := h.srv.Update(r.Context(), friend); err != nil {
+		respond.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+	respond.JSON(w, http.StatusOK, map[string]string{"result": "success"})
```
Comment (pthethanh) [inline internal/app/api/handler/friend/friend.go]: I think should return `"id": friend.ID

File: internal/app/friend/mongodb_repo.go [hunk 21]
```diff
@@ -28,11 +28,45 @@ func (r *MongoRepository) FindByID(ctx context.Context, id string) (*types.Frien
 	defer s.Close()
 	var friend *types.Friend
 	if err := r.collection(s).Find(bson.M{"id": id}).One(&friend); err != nil {
-		return nil, errors.Wrap(err, "failed to find the given friend from database")
+		return nil, errors.Wrap(err, "Failed to find the given friend from database")
```
Comment (pthethanh) [inline internal/app/friend/mongodb_repo.go]: Error can be wrapped into another error hence should not be started with CAP. Same comment for others

File: internal/app/types/friend.go [hunk 22]
```diff
@@ -2,6 +2,16 @@ package types
 
 // Friend hold information of a friend
 type Friend struct {
-	ID   string `json:"id" bson:"id"`
-	Name string `json:"name"`
-}
+	ID    			string  `json:"id,omitempty" bson:"id"`
```
Comment (pthethanh) [inline internal/app/types/friend.go]: bson should be _id

File: internal/app/api/handler/friend/friend.go (hunk 18 above)
Comment (sangvu) [reply internal/app/api/handler/friend/friend.go]: Fixed

File: internal/app/api/handler/friend/friend.go (hunk 19 above)
Comment (sangvu) [reply internal/app/api/handler/friend/friend.go]: Fixed

File: internal/app/api/handler/friend/friend.go (hunk 20 above)
Comment (sangvu) [reply internal/app/api/handler/friend/friend.go]: Fixed

File: internal/app/friend/mongodb_repo.go (hunk 21 above)
Comment (sangvu) [reply internal/app/friend/mongodb_repo.go]: Fixed

File: internal/app/types/friend.go (hunk 22 above)
Comment (sangvu) [reply internal/app/types/friend.go]: Fixed

**Review by pthethanh — APPROVED** [review]

**Pull request merged**